plugins/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureStorageService.java

```diff
@@ -103,9 +121,25 @@ private static CloudBlobClient createClient(AzureStorageSettings azureStorageSet
         return CloudStorageAccount.parse(connectionString).createCloudBlobClient();
     }
 
-    private static OperationContext buildOperationContext(AzureStorageSettings azureStorageSettings) {
+    private static OperationContext buildOperationContext(AzureStorageSettings azureStorageSettings,
+                                                          RequestMetricCollector metricCollector) {
         final OperationContext context = new OperationContext();
         context.setProxy(azureStorageSettings.getProxy());
+        if (metricCollector.isEnabled()) {
+            context.getRequestCompletedEventHandler().addListener(new StorageEvent<>() {
```

Comment:
Maybe we should just keep this logic in `AzureBlobStore` and simply add the listener to the requests that we meter just like we do in S3 (we can just add a utility for it `contex` + `RequestMetricCollector` in -> `context` out). Having this here with two separate methods for getting the client tuple is much harder to follow isn't it?